components/codec/src/number.rs

```diff
@@ -482,19 +484,58 @@ impl NumberCodec {
         let (uv, decoded_bytes) = Self::try_decode_var_u64(buf)?;
         let v = uv >> 1;
         if uv & 1 == 0 {
+            // no need for likely/unlikely here
             Ok((v as i64, decoded_bytes))
         } else {
             Ok((!v as i64, decoded_bytes))
         }
     }
+
+    /// Gets the length of the first encoded VarInt in the given buffer. If the buffer is not
+    /// complete, the length of buffer will be returned.
+    ///
+    /// This function is more efficient when `buf.len() >= 10`.
+    #[inline]
+    pub fn get_first_encoded_var_int_len(buf: &[u8]) -> usize {
+        unsafe {
+            let mut ptr = buf.as_ptr();
+            let len = buf.len();
+            if likely(len >= MAX_VARINT64_LENGTH) {
+                // Fast path
+                // Compiler will do loop unrolling for us.
+                for i in 1..=9 {
+                    if *ptr < 0x80 {
+                        return i;
+                    }
+                    ptr = ptr.add(1);
+                }
+                10
+            } else {
+                let ptr_end = buf.as_ptr().add(len);
+                // Slow path
+                while ptr != ptr_end && *ptr >= 0x80 {
+                    ptr = ptr.add(1);
+                }
+                // When we got here, we are either `ptr == ptr_end`, or `*ptr < 0x80`.
```

Comment:
The assert is not always true, e.g `[0xff,0xff,0xff,0xff,0xff,0xff,0xff,0xff,0x7f]`(9 bytes) both `ptr == ptr_end` and `*ptr < 0x80` are true.

Reply:
```
[0xff,0xff,0xff,0xff,0xff,0xff,0xff,0xff,0x7f]
                                             ^ptr_end
```

this is ptr_end. Your case is only `*ptr < 0x80`